src/gen-mapping.ts

```diff
@@ -336,6 +350,98 @@ export class GenMapping {
           : [genColumn, sourcesIndex, sourceLine, sourceColumn],
       );
     };
+
+    applySourceMap = (
+      map: GenMapping,
+      consumer: TraceMap,
+      rawSourceFile?: string,
+      sourceMapPath?: string,
+    ) => {
+      let sourceFile = rawSourceFile;
+
+      if (sourceFile == null) {
+        if (consumer.file == null) {
+          throw new Error(
+            'applySourceMap requires either an explicit source file, ' +
+              'or the source map\'s "file" property. Both were omitted.',
+          );
+        }
+        sourceFile = consumer.file;
+      }
+
+      const sourceRoot = map.sourceRoot;
+
+      // Make "sourceFile" relative if an absolute Url is passed.
+      if (sourceRoot != null) {
+        sourceFile = relative(sourceRoot, sourceFile);
+      }
+
+      const sourceIndex = get(map._sources, sourceFile);
+
+      // If the applied source map replaces a source entirely
+      // it's better to start with a fresh set of source, sourceContent and names
+      const newSources = new SetArray();
+      const newSourceContents: (string | null)[] = [];
+      const newNames = new SetArray();
```

Comment:
Since it's not possible to remove sources or names, it seems better to start from a fresh sources/sourceContents and names.

However I don't know if it's better to
* copy the existing objects, clear them and modify them in place (not what's done here)
* create new objects on the side, modify them and replace at the end (what's done here)